### PR TITLE
[Polkadot] Allow network value > 64 for SS58Address

### LINKED ISSUE #2050: [Polkadot] Allow network value > 64 for `SS58Address`
From: https://docs.substrate.io/v3/advanced/ss58/#address-type

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

In Substrate address implementation SS58Address, support network values larger than 64, with 2-bytes network value prefix.
Fixes #2050

For bit manipulation, following were used as source:
- https://github.com/paritytech/substrate/blob/master/primitives/core/src/crypto.rs
- https://docs.substrate.io/v3/advanced/ss58/#address-type
though the second doc is a bit misleading.

## Testing instructions

Unit tests.

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

### CODE REVIEW

File: src/SS58Address.h [hunk 1]
```diff
@@ -85,8 +91,45 @@ class SS58Address {
...
-        return Data(bytes.begin() + 1, bytes.end());
+    Data keyBytes() const {
+        byte networkSize;
+        uint64_t networkTemp;
+        decodeNetwork(bytes, networkSize, networkTemp);
+        return Data(bytes.begin() + networkSize, bytes.end());
+    }
+
+    // Return true and the network size (1 or 2) and network if input is valid
+    static bool decodeNetwork(const Data& data, byte& networkSize, uint64_t& network) {
```
Comment (hewigovens) [inline src/SS58Address.h]: Let's move `decodeNetwork` and `encodeNetwork` to `ScaleCodec.h`?

File: src/SS58Address.h [hunk 2]
```diff
@@ -85,8 +91,45 @@ class SS58Address {
...
+        networkSize = 0;
+        network = 0;
+        if (data.size() >= 1 && data[0] < 0x40) { // 0 -- 63
```
Comment (hewigovens) [inline src/SS58Address.h]: let's add constants for 0x40, 0x80

File: tests/Polkadot/AddressTests.cpp [hunk 3]
```diff
@@ -46,6 +46,6 @@ TEST(PolkadotAddress, FromPublicKey) {
 }
 
 TEST(PolkadotAddress, FromString) {
-    auto address = Address("15KRsCq9LLNmCxNFhGk55s5bEyazKefunDxUH24GFZwsTxyu");
-    ASSERT_EQ(address.string(), "15KRsCq9LLNmCxNFhGk55s5bEyazKefunDxUH24GFZwsTxyu");
+        auto address = Address("15KRsCq9LLNmCxNFhGk55s5bEyazKefunDxUH24GFZwsTxyu");
+        ASSERT_EQ(address.string(), "15KRsCq9LLNmCxNFhGk55s5bEyazKefunDxUH24GFZwsTxyu");
```
Comment (hewigovens) [inline tests/Polkadot/AddressTests.cpp:50]:
```suggestion
    auto address = Address("15KRsCq9LLNmCxNFhGk55s5bEyazKefunDxUH24GFZwsTxyu");
    ASSERT_EQ(address.string(), "15KRsCq9LLNmCxNFhGk55s5bEyazKefunDxUH24GFZwsTxyu");
```

**Review by hewigovens — COMMENTED** [review]
`Polkadot/SS58Address.h` is not checked in, why move it btw?

File: tests/Polkadot/AddressTests.cpp (hunk 3 above)
Comment (catenocrypt) [reply tests/Polkadot/AddressTests.cpp:50]: Only one half of the 'git mv' got checked in, fixed.
I moved SS58Address from generic `src` to Polkadot, as it is used only by Polkadot and Kusama.

File: src/SS58Address.h (hunk 2 above)
Comment (catenocrypt) [reply src/SS58Address.h]: +1

File: src/SS58Address.h (hunk 1 above)
Comment (catenocrypt) [reply src/SS58Address.h]: These are a bit different than the encoding functions in ScaleCodec, and used only by the address.

File: src/Polkadot/SS58Address.h [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        if (!isValid(string, network)) {
+            throw std::invalid_argument("Invalid address string");
+        }
+        const auto decoded = Base58::bitcoin.decode(string);
+        bytes.resize(decoded.size() - checksumSize);
+        std::copy(decoded.begin(), decoded.end() - checksumSize, bytes.begin());
+    }
+
+    /// Initializes an address with a public key and network.
+    SS58Address(const PublicKey& publicKey, int32_t network) {
```
Comment (hewigovens) [inline src/Polkadot/SS58Address.h:80]: Let's unify all `network` types, should `int32_t` work other places?
```suggestion
    SS58Address(const PublicKey& publicKey, uint64_t network) {
```

File: tests/Polkadot/AddressTests.cpp (hunk 3 above)
Comment (hewigovens) [reply tests/Polkadot/AddressTests.cpp:50]: These indentions still looks weird

File: tests/Polkadot/SS58AddressTests.cpp [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+#include <gtest/gtest.h>
+#include <map>
+
+using namespace TW;
+
+TEST(SS58Address, IsValid) {
+    EXPECT_TRUE(SS58Address::isValid("15KRsCq9LLNmCxNFhGk55s5bEyazKefunDxUH24GFZwsTxyu", 0));
+    
+    EXPECT_TRUE(SS58Address::isValid("VNpyemgLHFKxYhWKzSKM5SoGYyAqtBSdyRaQ3DcNEsT9jpVnx", 110));
```
Comment (hewigovens) [inline tests/Polkadot/SS58AddressTests.cpp:20]: Are those networks really existed? where can we cross verify our address test cases?

**Review by hewigovens — APPROVED** [review]

File: src/Polkadot/SS58Address.h (hunk 4 above)
Comment (catenocrypt) [reply src/Polkadot/SS58Address.h]: Fixed, uint64_t was a mistake

File: tests/Polkadot/SS58AddressTests.cpp (hunk 5 above)
Comment (catenocrypt) [reply tests/Polkadot/SS58AddressTests.cpp]: Good point, I did not find official test to cross-check

Comment (catenocrypt) [reply tests/Polkadot/SS58AddressTests.cpp]: Cross-checked with Substrate implementation from https://github.com/paritytech/substrate/blob/master/primitives/core/src/crypto.rs

**Review by FitzLu — APPROVED** [review]

**Pull request merged**